### PR TITLE
Time-dependent GO adaptation demo and time-dependent constants

### LINKED ISSUE #28: Time-dependent goal-oriented mesh adaptation demo
Parent: pyroteus/pyroteus#94.

### PR DESCRIPTION
**Pull request by ddundo**
Closes #28.

In this PR I add a bubble shear goal-oriented adaptation demo. It involves a time-varying velocity field which is prescribed and not solved for. So to ease into the GO demo, I also add a simple `solve_forward` bubble shear demo to introduce the idea of "time-dependent" constants.

Related to #73 discussion: I added a check in `GoalOrientedMeshSequence.indicate_errors` to see whether the user flagged in their `get_form` that there are time-dependent constants. In that case, I recompile the form for each exported timestep, which should update these fields. By "flag" I mean that the form function contains the kwarg `err_ind_time`, i.e.: `def form(index, fields, err_ind_time=None)`. I am not fully happy with this so I also didn't fully tidy up the go_mesh_seq.py code, as I hope to get feedback from you on how to improve this - but here is my reasoning behind it:

1. This does not change the interface when we do not have time-dependent constants,
2. We do not need to recompile the form for every timestep when we are not inside `indicate_errors()`. Instead we can (and should) update the time-dependent constants in `get_solver`.
3. The name `err_ind_time` is unique enough to be sure that the user won't randomly choose it.
4. This kind of "flag" is necessary to handle the 2nd part of Stephan's comment here https://github.com/pyroteus/goalie/issues/55#issuecomment-1702280655 

Overall, I think it's a good solution but I am happy to keep #73 open and explore other ideas.

I will make 2 comments in the code myself where things are left undone.

And I tried to keep the demos cheap (to prevent the testing suite taking too long) so the results could be better, but they still look good I'd say!
image

### CODE REVIEW

File: demos/bubble_shear.py [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+# interest.
+#
+# First, we import Firedrake and Goalie. ::
+
+from firedrake import *
+from goalie import *
+
+# We begin by definining the background velocity field :math:`\mathbf{u}(x, y, t)`.
+# Specifically, we choose the velocity field that is periodic in time, such as
+# in TODO: cite, and is given by
```
Comment (ddundo) [inline demos/bubble_shear.py:27]: I added the citation in demo_references.bib but not sure how to make the citation like in other demos.

File: goalie/go_mesh_seq.py [hunk 2]
```diff
@@ -150,18 +155,26 @@ def indicate_errors(
...
-                    "The function defined by get_form should return a dictionary"
-                    f", not type '{type(forms)}'."
-                )
+            if not timedep_const:
+                forms = enriched_mesh_seq.form(i, mapping)
+                if not isinstance(forms, dict):
+                    raise TypeError(
+                        "The function defined by get_form should return a dictionary"
+                        f", not type '{type(forms)}'."
+                    )
```
Comment (ddundo) [inline goalie/go_mesh_seq.py:164]: After my changes, this check is not being done if `timedep_const=True`. But I thought that we could find a better place for this to live anyway.

**Comment by ddundo** [comment]
I cancelled the workflow because it looked like the GO demo was taking a while (takes about 2-3 minutes for me locally). We should cut it down even further (e.g. use only one fp iteration, decrease `end_time`...).

File: demos/bubble_shear.py [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+# conditions for the concentration, so in this demo we solve
+#
+# .. math::
+#   \begin{array}{rl}
+#       \frac{\partial c}{\partial t} + \nabla\cdot(\mathbf{u}c)=0& \text{in}\:\Omega\\
+#       c=0 & \text{on}\:\partial\Omega,\\
+#       c=c_0(x,y) & \text{at}\:t=0,
+#   \end{array},
+#
+# where :math:`c=c(x,y,t)` is the sought tracer concentration, :math:`\mathbf{u}=\mathbf{u}(x,y,t)` is the
```
Comment (jwallwork23) [inline demos/bubble_shear.py:16]: Please keep lines to 89 characters, except in the `.. math::` blocks.

File: demos/bubble_shear.py [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+test = os.environ.get("GOALIE_REGRESSION_TEST") is not None
+n = 50 if not test else 10
+dt = 0.01 if not test else 0.025
+
+num_subintervals = 2
+meshes = [UnitSquareMesh(n, n) for _ in range(num_subintervals)]
+end_time = period / 2
+time_partition = TimePartition(
+    end_time,
+    len(meshes),
```
Comment (jwallwork23) [inline demos/bubble_shear.py:193]: Would prefer `num_subintervals` here.

File: demos/bubble_shear.py [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+# periodicity of the velocity field :math:`\mathbf{u}`, which we chose to be an odd
+# function of time. This means that the velocity field reverses direction at :math:`t=T/2`,
+# and so does the deformation of the tracer bubble, which returns to its initial shape.
+#
+# However, while the tracer bubble at :math:`t=T/2` does resemble a very blurry circular
+# shape, it is far from matching the initial condition. This "bluriness" is the result of
+# adding the SUPG stabilisation to the weak form, which adds numerical diffusion. This
+# diffusion is necessary for numerical stability and for preventing oscillations, but it
+# also makes the solution irreversible. The amount of difussion added is related to the
+# grid Peclet number :math:`Pe = U\,h/2D`: the coarser the mesh is, the more diffusion
```
Comment (jwallwork23) [inline demos/bubble_shear.py:231]: Minor point: "Péclet number :math:`\mathrm{Pe} = U\,h/2D`"

File: demos/bubble_shear.py [hunk 6]
```diff
@@ -0,0 +1,237 @@
+# Advection of a tracer bubble in a non-uniform shear flow
+# ===================================================
+
+# In this demo, we consider the advection of a scalar concentration field, similarly to
+# our previous solid body rotation demo. However, we now consider a background velocity
```
[the remark this answers is not in the document]
Comment (jwallwork23) [reply demos/bubble_shear.py:5]: This demo will follow on from the Gray-Scott demo so please provide references between this demo and that one so that the demos can be read through in order.

File: goalie/go_mesh_seq.py (hunk 2 above)
Comment (jwallwork23) [reply goalie/go_mesh_seq.py:164]: Surely `forms` needs to be defined in the `timedep_const` case? Did you have an idea of the 'better place' to put this?

File: demos/bubble_shear-goal.py [hunk 7]
```diff
@@ -0,0 +1,313 @@
...
+# the :meth:`get_form` method is called from within :class:`GoalOrientedMeshSeq`
+# while computing error indicators. There, only those fields that are passed to
+# :class:`GoalOrientedMeshSeq` are passed to :meth:`get_form`. This means that the velocity
+# field would not be updated in time. In such a case, we will manually compute the
+# velocity field using the current simulation time, which will be passed automatically
+# as an :arg:`err_ind_time` kwarg in :meth:`GoalOrientedMeshSeq.indicate_errors`. ::
+
+
+def get_form(mesh_seq):
+    def form(index, form_fields, err_ind_time=None):
```
Comment (jwallwork23) [inline demos/bubble_shear-goal.py:117]: I wonder if we should just add `time` as an argument to `form` in general? For problems where the form doesn't vary with time, it will just be ignored. IMO it would be better to only pass solution fields (and their lagged counterparts) through the second argument and always determine other variables based on the current time. I guess this would imply some refactoring, including the other demos.

It seems odd to reference error indicators in the form expression.

File: demos/bubble_shear-goal.py [hunk 8]
```diff
@@ -0,0 +1,313 @@
...
+            subinterval_metric.intersect(metric)
+        metrics.append(metric)
+
+    # Normalise the metrics in space and time
+    space_time_normalise(metrics, mesh_seq.time_partition, mp)
+
+    complexities = np.zeros(len(mesh_seq))
+    for i, metric in enumerate(metrics):
+        complexities[i] = metric.complexity()
+        mesh_seq[i] = adapt(mesh_seq[i], metric)
```
Comment (jwallwork23) [inline demos/bubble_shear-goal.py:220]: Side note: this makes me think about #122.

File: demos/bubble_shear-goal.py [hunk 9]
```diff
@@ -0,0 +1,313 @@
...
+n = 50 if not test else 5
+dt = 0.01 if not test else 0.05
+maxiter = 2 if not test else 1  # maximum number of fixed point iterations
+
+num_subintervals = 6
+meshes = [UnitSquareMesh(n, n) for _ in range(num_subintervals)]
+end_time = period / 2
+time_partition = TimePartition(
+    end_time,
+    len(meshes),
```
Comment (jwallwork23) [inline demos/bubble_shear-goal.py:251]: Again, please use `num_subintervals`.

File: demos/bubble_shear-goal.py [hunk 10]
```diff
@@ -0,0 +1,313 @@
...
+    # plot the solution at the final subinterval timestep
+    time = time_partition.subintervals[i][-1]
+    tripcolor(solutions["c"]["forward"][i][-1], axes=ax, cmap="coolwarm")
+    triplot(msq[i], axes=ax, interior_kw={"linewidth": 0.08})
+    ax.annotate(f"t={time:.2f}", (0.05, 0.05), color="white")
+
+fig.tight_layout(pad=0.3)
+fig.savefig("bubble_shear-goal_final_meshes.jpg", dpi=300, bbox_inches="tight")
+
+# .. figure:: bubble_shear-goal_final_meshes.jpg
```
Comment (jwallwork23) [inline demos/bubble_shear-goal.py:294]: Could you follow the naming pattern and have this file and its plots use `bubble_shear-goal_oriented*`?

**Review by jwallwork23 — CHANGES_REQUESTED** [review]
This is really great work, thanks so much @ddundo!

I have various minor comments. My main comment is the one suggesting that we actually adopt `time` as a third (compulsory) argument to `form`. Interested to hear what you think about this.

File: demos/bubble_shear-goal.py (hunk 7 above)
Comment (jwallwork23) [reply demos/bubble_shear-goal.py:117]: If you do end up making this change then it would be handy to fix #140 while you're at it if you don't mind.

File: demos/bubble_shear-goal.py (hunk 9 above)
Comment (ddundo) [reply demos/bubble_shear-goal.py]: Fixed in 262c78a51a99714a4794b55259edf6fc5aea3567.

File: demos/bubble_shear-goal.py (hunk 10 above)
Comment (ddundo) [reply demos/bubble_shear-goal.py]: Fixed in 262c78a51a99714a4794b55259edf6fc5aea3567.

File: demos/bubble_shear-goal.py [hunk 11]
```diff
@@ -0,0 +1,313 @@
+# Goal-oriented mesh adaptation for a 2D time-dependent problem
+# ===================================================
```
[the remark this answers is not in the document]
Comment (ddundo) [reply demos/bubble_shear-goal.py]: Fixed in 262c78a51a99714a4794b55259edf6fc5aea3567.

File: demos/demo_references.bib [hunk 12]
```diff
@@ -16,3 +16,10 @@ @Book{Hundsdorfer:2003
   year={2003},
   publisher={Springer}
 }
+
+@article{barral2016anisotropic,
```
[the remark this answers is not in the document]
Comment (ddundo) [reply demos/demo_references.bib]: Fixed in 262c78a51a99714a4794b55259edf6fc5aea3567.

File: demos/bubble_shear.py [hunk 13]
```diff
@@ -0,0 +1,237 @@
+# Advection of a tracer bubble in a non-uniform shear flow
+# ===================================================
```
[the remark this answers is not in the document]
Comment (ddundo) [reply demos/bubble_shear.py]: Fixed in 262c78a51a99714a4794b55259edf6fc5aea3567.

File: demos/bubble_shear.py (hunk 5 above)
Comment (ddundo) [reply demos/bubble_shear.py]: Fixed in 262c78a51a99714a4794b55259edf6fc5aea3567.

**Comment by ddundo** [comment]
Thanks a lot for reviewing this @jwallwork23! I addressed the minor comments in the new commit. 

> My main comment is the one suggesting that we actually adopt `time` as a third (compulsory) argument to `form`. Interested to hear what you think about this.

I might be missing an obvious solution... but if we make this change, I don't see how we would separate out the problems where we do and do not have these time-dependent constants. So in `error_indicators` we would always pass `time` to `get_form`, which would mean recompiling the form unnecessarily in all other demos we have. That's what I meant when I said that the kwarg `err_ind_time` flags that we have these fields, so we need to call `get_form` in order to update them since they're not stored in the solution dictionaries. Do you have a suggestion how to deal with this please?

File: demos/bubble_shear-goal.py (hunk 7 above)
Comment (ddundo) [reply demos/bubble_shear-goal.py]: Sorry @acse-ej321, I just saw this! Please see the new comment I left explaining my reasoning here. I agree that this suggestion is much nicer but I think it's also much more inefficient - unless I'm missing something! Please correct me if I'm wrong :)

File: demos/bubble_shear.py (hunk 3 above)
Comment (ddundo) [reply demos/bubble_shear.py]: Fixed in 262c78a51a99714a4794b55259edf6fc5aea3567 and I'll keep an eye on it if I make further changes! Thanks :)

File: demos/bubble_shear.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (ddundo) [reply demos/bubble_shear.py]: Fixed in 262c78a51a99714a4794b55259edf6fc5aea3567.

**Comment by jwallwork23** [comment]
Good point. I guess we need `get_form` to produce an `update_form` method, which could get stashed on the `MeshSeq`? Then - if it exists - it is called at each timestep?

**Comment by acse-ej321** [comment]
@ddundo @jwallwork23  - I don't think you would need to go to that extreme in changing the `def solver` function. For me, the current setup works, 

where I pass `t` as a parameter to `get_form`:
```
 def get_form(self):
        def form(index, solutions,t,field="u"):
```
and define it in `def solver` outside the `update_forcing` or `t` stepping loop as a `firedrake.Constant` such that in the loop the `fd.Constant` is updated. 
```
 def get_solver(self):
        def solver(index, ic):

            # Time integrate from t_start to t_end
            t_start, t_end = self.subintervals[index]
            dt = self.time_partition.timesteps[index]
            t = fd.Constant(t_start)
            t_= t_start

           # assemble the form outside the loop
           F = self.form(index, {"u": (u, u_)},t, field="u")["u"]

          while t_ < t_end - 1.0e-05:
                # solve
                fd.solve(F == 0, u,bcs=bcs, ad_block_tag="u")
                # reassign prior
                u_.assign(u)
                # update time stepping constant for loop
                t_ += dt
                # update firedrake.Constant time in form with 'assign'
                t.assign(t_)
```

Firedrake will automatically assemble the `form` with the updated `Constant` as far as I am aware, as long as you `assign it - so there would be no need to nest the assembly of the form in the `t`-stepping loop for that intention. That is not to say an update form might be more elegant?

**Comment by ddundo** [comment]
@acse-ej321 I may be misunderstanding (please correct me!), but if you take a look at L125-L180 in https://github.com/firedrakeproject/firedrake/blob/master/firedrake/solving.py, when you do `fd.solve(F == 0, u,bcs=bcs, ad_block_tag="u")`, firedrake assembles the `NonlinearVariationalProblem`. So when we do it inside the `while` loop, it does this for every timestep, which is very inefficient. So it's better to do it outside of the `while` loop. I get about 10x speed increase when I assemble it outside the loop.

**Comment by ddundo** [comment]
>I don't see how we would separate out the problems where we do and do not have these time-dependent constants. So in `error_indicators` we would always pass `time` to `get_form`, which would mean recompiling the form unnecessarily in all other demos we have.

I went to test how expensive this is. I called `get_form` 10 000 times and it took only 21 seconds! So I think we can sacrifice this bit of inefficiency in `indicate_errors` and always recompile the form? The extra cost of this is dwarfed by everything else in `indicate_errors` :)

> Good point. I guess we need `get_form` to produce an `update_form` method, which could get stashed on the `MeshSeq`? Then - if it exists - it is called at each timestep?

Was this referring to calling `update_form` inside the `solver` or `indicate_errors`? In `get_solver` it doesn't really matter for goalie, right? Users can do whatever they want there anyway. And I think we should keep it like this - I don't see a point in imposing some restrictions here.

So in summary: now I don't see a downside to making `time` a compulsory argument in `form`. What do you think @acse-ej321  and @jwallwork23? :)

**Comment by jwallwork23** [comment]
Yes, in general we should create the NVP and repeatedly its solve method rather than calling the `solve` driver function - I had just used that in the preliminary demos because its more intuitive for beginners.

**Comment by jwallwork23** [comment]
Okay how about this: we introduce a `time` attribute for `MeshSeq`, which is a list containing a Function from R-space for each subinterval, initialised to the start time of that subinterval. Inside the solver, we get this time and use it in the loop. Since it's a Function rather than a Constant, it will still support `time += dt`. The same `time[subinterval]` should be used inside the form, which would mean that updating the time in the solver loop would automatically update the form, too. Similar changes presumably required for the error indication code.

Some comments:
* Probably best to have the time attribute private and provide a public `@property` that grabs it.
* The time values would need to be reset to the start times of the corresponding subintervals before solving.

**Comment by ddundo** [comment]
Actually sorry, I'm not sure I see how that would allow us to address the last paragraph of https://github.com/pyroteus/goalie/issues/55#issuecomment-1702280655.

That is, in the gray_scott_split.py demo, in `solver` we have:
```
while t < t_end - 0.5 * dt:
    nlvs_a.solve()
    nlvs_b.solve()
```
which means that in `form` we have to use the values of `b` and `b_` from the previous timestep when compiling `F_a`:
```
# no changes needed for F_b

if inside_error_indicators:
    b = b_             # b from the previous timestep
    b_ = b_old_old     # b_ from the previous timestep
F_a = ...
```

So in your suggestion @jwallwork23, what would this `if inside_error_indicators` condition actually be?

E.g. if we had `def form(index, solutions, time=None)` then this condition could be `if time is not None`, since we do not pass `time` from `solver` but pass it from `indicate_errors`.

**Comment by jwallwork23** [comment]
Sorry @ddundo I'm confused.
* I am no longer suggesting passing `time` as a kwarg. Instead I'm suggesting it to be an attribute of the MeshSeq, with one time Function for each subinterval.
* I don't like this conditional `if inside_indicate_errors`. Can we not avoid this?
* Let's not get ahead of ourselves - we can think about how to address other PRs later.

**Comment by ddundo** [comment]
Yeah, sorry about that... I realised I jumped ahead after I wrote it. I will have more time over the weekend so I will code up your suggestion and post an update :) I'm just trying to avoid refactoring in the future. Thanks again both for the input!

**Comment by ddundo** [comment]
I was trying out @jwallwork23's suggestion now and I don't really see a nice way to make it work... could you please tell me if I am missing something obvious?

So just considering the basic `solve_forward` for now - nothing goal-oriented here... if I understood correctly, @jwallwork23 suggested to:
- introduce `time = MeshSeq.time`
- not pass velocity fields `u, u_` from solver to `get_form` but rather use `time` to determine them inside `get_form`

So this would look something like this:
```
def get_form(mesh_seq):
    def form(index, solutions):
        Q = mesh_seq.function_spaces["c"][index]
        V = VectorFunctionSpace(mesh_seq[index], "CG", 1)
        R = FunctionSpace(mesh_seq[index], "R", 0)
        dt = Function(R).assign(mesh_seq.time_partition.timesteps[index])
        theta = Function(R).assign(0.5)  # Crank-Nicolson implicitness parameter

        c, c_ = solutions["c"]

        # Velocity
        x, y = SpatialCoordinate(mesh_seq[index])
        u = Function(V).interpolate(velocity_expression(x, y, mesh_seq.time[index]))
        u_ = Function(V).interpolate(velocity_expression(x, y, mesh_seq.time[index] - dt))

        # SUPG stabilisation parameter
        D = Function(R).assign(0.1)  # diffusivity coefficient
        h = CellSize(mesh_seq[index])  # mesh cell size
        U = sqrt(dot(u, u))  # velocity magnitude
        tau = 0.5 * h / U
        tau = min_value(tau, U * h / (6 * D))

        phi = TestFunction(Q)
        phi += tau * dot(u, grad(phi))

        a = c * phi * dx + dt * theta * dot(u, grad(c)) * phi * dx
        L = c_ * phi * dx - dt * (1 - theta) * dot(u_, grad(c_)) * phi * dx
        F = a - L

        return {"c": F}
    return form


def get_solver(mesh_seq):
    def solver(index, ic):
        tp = mesh_seq.time_partition
        dt = tp.timesteps[index]
        num_timesteps = tp.num_timesteps_per_subinterval[index]

        t = mesh_seq.time[index]

        # Initialise the concentration fields
        Q = mesh_seq.function_spaces["c"][index]
        c = Function(Q, name="c")
        c_ = Function(Q, name="c_old").assign(ic["c"])

        F = mesh_seq.form(index, {"c": (c, c_)})["c"]
        nlvp = NonlinearVariationalProblem(F, c, bcs=mesh_seq.bcs(index))
        nlvs = NonlinearVariationalSolver(nlvp, ad_block_tag="c")

        # Time integrate
        for _ in range(num_timesteps):
            # TODO: Update u and u_ without reassembling nlvp?

            # solve the advection equation
            nlvs.solve()

            # update the 'lagged' concentration
            c_.assign(c)
            t += dt

        return {"c": c}
    return solver
```

I don't see how we can update the velocity fields inside the timestepping loop without calling `get_form` again, which would require assembling NLVP again - which we want to avoid. Do you see a way to make it work please?

An (ugly) option I see is to have `get_form` return `(u, u_)`. In that case we can then do something like this:
```
def get_form(mesh_seq):
    def form(index, solutions):
        # all the same as above

        extra_fields = {"u": (u, u_)}

        return {"c": F}, extra_fields 
    return form


def get_solver(mesh_seq):
    def solver(index, ic):
        # same as above up to this point

        form, extra_fields = mesh_seq.form(index, {"c": (c, c_)})

        F = form["c"]
        nlvp = NonlinearVariationalProblem(F, c, bcs=mesh_seq.bcs(index))
        nlvs = NonlinearVariationalSolver(nlvp, ad_block_tag="c")

        u, u_ = extra_fields["u"]

        # Time integrate
        for _ in range(num_timesteps):
            # update the velocity field
            u.interpolate(velocity_expression(x, y, mesh_seq.time[index]))

            # solve the advection equation
            nlvs.solve()

            # update the 'lagged' concentration and velocity field
            c_.assign(c)
            u_.assign(u)
            t += dt

        return {"c": c}
    return solver
```

This is ugly but it would also easily solve the problem in `indicate_errors` where the only change necessary would be adding this one line
```
forms = enriched_mesh_seq.form(i, mapping)
forms = forms if len(forms) == 1 else forms[0]
```
So this would solve all issues in this particular PR.

But this doesn't help us at all with the issue in gray_scott_split.py that I mentioned above... So I think it would require refactoring in the future.

File: goalie/go_mesh_seq.py (hunk 2 above)
Comment (ddundo) [reply goalie/go_mesh_seq.py:164]: Yup! See #169 :)

File: goalie/go_mesh_seq.py [hunk 14]
```diff
@@ -87,54 +91,89 @@ def get_enriched_mesh_seq(
...
+        self.solve_adjoint(**solver_kwargs)
+        enriched_mesh_seq.solve_adjoint(**solver_kwargs)
+
+        tp = self.time_partition
+        # check whether get_form contains a kwarg indicating time-dependent constants
+        timedep_const = "err_ind_time" in signature(enriched_mesh_seq.form).parameters
+
+        tp = self.time_partition
+        # check whether get_form contains a kwarg indicating time-dependent constants
+        timedep_const = "err_ind_time" in signature(enriched_mesh_seq.form).parameters
```
Comment (jwallwork23) [inline goalie/go_mesh_seq.py:168]: Not sure what happened here with the duplication. But - in either case - we can get rid of the `signature` stuff now that `err_ind_time` has been dropped.

**Review by jwallwork23 — CHANGES_REQUESTED** [review]
@ddundo Would you mind addressing my comment and rebasing on top of `main`?

File: goalie/go_mesh_seq.py (hunk 14 above)
Comment (ddundo) [reply goalie/go_mesh_seq.py]: Done in 29e959f5b82f28c496433347c3492008a868354c

**Comment by ddundo** [comment]
> @ddundo Would you mind addressing my comment and rebasing on top of `main`?

I've done it but I didn't update the bubble shear demos. I'd prefer to wait for #164 to be merged before making further changes, since that properly takes care of #55, which overlaps with changes in this PR. So I think it would make sense to prioritise that PR before this one for that reason :)

File: goalie/go_mesh_seq.py [hunk 15]
```diff
@@ -130,6 +131,10 @@ def indicate_errors(
         self.solve_adjoint(**adj_kwargs)
         enriched_mesh_seq.solve_adjoint(**adj_kwargs)
 
+        tp = self.time_partition
+        # check whether get_form contains a kwarg indicating time-dependent constants
+        timedep_const = "err_ind_time" in signature(enriched_mesh_seq.form).parameters
```
[the remark this answers is not in the document]
Comment (ddundo) [reply goalie/go_mesh_seq.py]: Resolving this since it's no longer relevant :)

File: goalie/go_mesh_seq.py [hunk 16]
```diff
@@ -150,18 +155,26 @@ def indicate_errors(
...
                 # Loop over each timestep
                 for j in range(len(self.solutions[f]["forward"][i])):
                     # Update fields
+                    if timedep_const:
+                        # recompile the form with updated time-dependent constants
+                        time = (
+                            tp.subintervals[i][0]
+                            + (j + 1) * tp.timesteps[i] * tp.num_timesteps_per_export[i]
+                        )
+                        forms = enriched_mesh_seq.form(i, mapping, err_ind_time=time)
```
[the remark this answers is not in the document]
Comment (ddundo) [reply goalie/go_mesh_seq.py]: Done :)

**Comment by ddundo** [comment]
I converted this to ready for review to draw more attention to it :) as I mentioned in the meeting (and as discussed in the long thread above), it remains to decide how to communicate these "time-dependent constants" within Goalie.

The two new demos fail - I haven't updated them yet to `yield` etc.

**Comment by ddundo** [comment]
Hi @jwallwork23 @stephankramer @acse-ej321, here is the summary of the issue :) I will also start by saying that the PR is longish because there are 2 demos, but it is enough to just look at one of them (either one) for the purposes of the discussion here

In this specific bubble shear case example, we have a background velocity field that we do not solve for - we just update it at each timestep. How goalie is currently set up, when we call `indicate_errors` we only update the prognostic fields that we solved for (and which we saved at each export timestep). We also need to update this velocity field at each export timestep in order to compute the error indicators correctly.

----------------

Here is the idea that I had how we could do this automatically, without requiring extra effort from the user.

We could identify these changing fields and extract them from the `coefficients` attribute of the variational form. I made a small gist to help demonstrate: https://gist.github.com/ddundo/92fc7d9fd24471a37c5a903ddd035554 - I labelled the important part towards the end. Here it's all inside the "user code", but it could all be done really nicely in the goalie code. 

So we could extract and save a copy of these fields at each exported timestep, before then using them in `indicate_errors`. This is potentially memory-intensive, so I also suggest that we do the following. At the moment, goalie is set up so that we solve the forward and adjoint problem over all subintervals, and only then do we call `indicate_errors`. I think that we should instead call `indicate_errors` after each individual subinterval. That way:
- we only need to store these changing fields for the export timesteps in the subinterval, rather than over the whole time interval
- similarly for `forward_old` and `adjoint_next` fields which we are now storing over the whole time interval: we could drop them after each call to `indicate_errors` and free up memory after each subinterval

**Comment by jwallwork23** [comment]
Apologies for the delay getting back to you @ddundo. This sounds like a great approach to me. It'd be preferable if you could break down the efforts into several small PRs, if possible. e.g., start by splitting up the `indicate_errors` call in one PR. That sound okay?

**Comment by ddundo** [comment]
Thanks @jwallwork23! I'll do it in steps as you suggested :)
